ENH: multi channels API, coloring and blending

Punzo opened the pull request:
Draft PR regarding https://github.com/OHIF/Viewers/issues/2199

Screenshot running the new blend example:
<img src="https://user-images.githubusercontent.com/7985338/109158986-5facbd80-7774-11eb-926a-d51c3ebebc49.png" width="512">

**NOTE: we color only monochrome channels.** 

Dev Notes: document https://docs.google.com/document/d/1LM457qPBG1JIlwCDQQOQ1XbRC0Rz9Ea5QmN9aHny1lw/edit

TO DO:
- [x] (1) fix flickering when compositing the channels. **NOTE**: the issue is generated when openLayer changes opacity if a layer is in "transition" or not (https://github.com/openlayers/openlayers/blob/main/src/ol/renderer/canvas/TileLayer.js#L476).
- [ ] (2) extend the API to control also the focal plane attribute (i.e., 3D channels). **NOTE**: we need data, we will wait ~1 month to get more official HTAN/IDC data. For the moment just ignore instances that have dimensionOrganizationType !== TILED_FULL
- [x] (3) update the blend example to use a server dcm4chee instance. **NOTE**. upload LUNG-1-LN data public on the dcmjs.server.org instance and cite https://www.synapse.org/#!Synapse:syn17865732/wiki/592782
- [ ] (4) use DICOM attributes to load/save the state. **NOTE**: we are waiting for David’s feedback regarding the advanced blending presentation state.
- [x] (5): for coloring monochrome channels we download the uncompressed binaries data (i.e. we use mediaType = 'application/octet-stream' and transferSyntaxUID = '1.2.840.10008.1.2.1'). For a large N of channels, this could be a substantial bottleneck. Check performance downloading uncompressed binary (octet-stream) vs downloading compressed (e.g.: jpeg) + decoding. **NOTE**: have a look at https://github.com/cornerstonejs/codecs

Punzo commented:
> * mediaType = 'application/octet-stream'

hackermd reviewed (changes requested):
Nice implementation @Punzo! I have a couple of comments. Let me try to summarize them here:

- **Examples**: I noticed that you added conditional statements to each example that check whether certain DICOM attributes are present. All of them are type 1 attributes and shall be present in any DICOM dataset. Why are these statements necessary? I would remove them from the examples. They should serve as a user guide on how to use the library and should thus be free of compliance violation workarounds.

- **WebGL rendering / blending**:
    * I think the code in `colorImageFrames.js` could be refactored into a `WebGLRenderer` class. This would improve readability and usability of the code in my opinion. The viewer could then instantiate a renderer and reuse the renderer object for rendering of individual frames.
    * Please add documentation to the public method(s), describing the parameters, but also the behavior and some of the expectations/assumptions. I particular, please add a high-level description of how the additive blending works.
    * Indicate which methods are "private" by renaming them using a leading underscore.

- **Viewer**:
    * Some code seems to be duplicated. Also, some parts could probably be factored out into separate functions. For example, the block for processing an individual channel has become very long.
    * I would suggest defining a class for managing the blending information (color, opacity, etc.), which can then be passed to the constructor of the viewer as well as to methods of the viewer that are indented to access or manipulate the presentation state (see next item).
    * Instead of using multiple methods to get/set individual presentation state properties (color, opacity, etc.), I would define just one method to get/set the presentation state. An app can then update several of the properties at once (and update the presentation state in one rendering operation).

pieper commented on examples/basic/index.html:
```diff
@@ -31,29 +34,38 @@
-          const imageType = metadata[0]["00080008"]["Value"];
-          if (imageType[2] === "VOLUME") {
-            return(metadata[0]);
+          const imageType = metadata[0]["00080008"]
```
As a rule I dislike code that uses dicom hex for any tags that are defined in the dictionary.  It makes the code less readable and is partly why dicom has such a bad reputation.  Can we change these?

hackermd replied:
Fully agree with @pieper. That's why I am suggesting to immediately format (sanitize) the metadata.

pieper replied:
In dcmjs, the pattern is to define a [subclass of Normalizer](https://github.com/dcmjs-org/dcmjs/blob/master/src/normalizers.js#L33-L57) for each SOPClass to catch and correct known deviations from the standard.  This can avoid spreading special case code all over the application logic.

pieper commented on examples/blend/index.html:
```diff
@@ -0,0 +1,136 @@
+      const channelInit = [channelOne, channelTwo, channelThree];
+
+      const blendingImageQuality = 0.8 // value between 0 and 1;
+      // Note: We are using retrieveRendered: false because dcm4chee does not yet
```
@swederik said that upgrading that server is on his todo list.  Did it already happen?

pieper commented:
> * I think the code in `colorImageFrames.js` could be refactored into a `WebGLRenderer` class.

+1 for abstracting out the webgl part.  [WebGPU](https://en.wikipedia.org/wiki/WebGPU) is not far off and should have advantages if it becomes widely available.

Punzo commented:
> Nice implementation @Punzo! I have a couple of comments. Let me try to summarize them here:
> 
> * **Examples**: I noticed that you added conditional statements to each example that check whether certain DICOM attributes are present. All of them are type 1 attributes and shall be present in any DICOM dataset. Why are these statements necessary? I would remove them from the examples. They should serve as a user guide on how to use the library and should thus be free of compliance violation workarounds.
> * **WebGL rendering / blending**:
>   
>   * I think the code in `colorImageFrames.js` could be refactored into a `WebGLRenderer` class. This would improve readability and usability of the code in my opinion. The viewer could then instantiate a renderer and reuse the renderer object for rendering of individual frames.
>   * Please add documentation to the public method(s), describing the parameters, but also the behavior and some of the expectations/assumptions. I particular, please add a high-level description of how the additive blending works.
>   * Indicate which methods are "private" by renaming them using a leading underscore.
> * **Viewer**:
>   
>   * Some code seems to be duplicated. Also, some parts could probably be factored out into separate functions. For example, the block for processing an individual channel has become very long.
>   * I would suggest defining a class for managing the blending information (color, opacity, etc.), which can then be passed to the constructor of the viewer as well as to methods of the viewer that are indented to access or manipulate the presentation state (see next item).
>   * Instead of using multiple methods to get/set individual presentation state properties (color, opacity, etc.), I would define just one method to get/set the presentation state. An app can then update several of the properties at once (and update the presentation state in one rendering operation).

thanks for the review, I will work on it in the next days!

hackermd commented:
> > * I think the code in `colorImageFrames.js` could be refactored into a `WebGLRenderer` class.
> 
> +1 for abstracting out the webgl part. [WebGPU](https://en.wikipedia.org/wiki/WebGPU) is not far off and should have advantages if it becomes widely available.

Good point. I like the idea of fully abstracting the WebGL rendering components. Maybe we should just call the class `RenderingEngine` or something similar.

Punzo commented:
Hi @hackermd @pieper, I have applied the review:
- complete refactoring/cleaning of the viewer constructor with new channel and blendingInformation classes.
- complete refactoring/cleaning of the offscreen render converting it in anew RenderingEngine class.
- this fix https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/47/commits/f080301b77d2cf61d241c99ea094258e2f25c433 cache the pixelData array at the end of a tile load. In this way we can reapply the offscren render when changing the blending parameters (opacity, clipping, color) withour downloading again the data.

----------------------------------------------------------------------
I also have looked at the DICOM presentation state (point(4) of the PR https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/47#issue-580068584). 

For each channel/layer we need the following parameters:
A) opticalPathIdentifer (string, used as id of the channel). 
B) Color (array[3], range [0,255])
C) Opacity (float, range [0,1]).
D) ContrastLimit (array[2], min and max. range [0,255]) : we can use 
E) Visibility (boolean)  
F) AddToMap (boolean). It is different from the visibility, it is to know if we have to add the layer that corresponds to the channel to the OpenLayer map object. 

the most promising DICOM module is the **Advanced Blending Presentation State**. Here some preliminary investigation for which I would like some feedback:

A) This is already a DICOM attribute - **OK**

B) unfortunately the **Advanced Blending Presentation State** (http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.11.34.html) is not ideal for this. Since we have a specific opacity value for each channel. While the DICOM attributes allows only one opacity value for N inputs (see Blending Mode = EQUAL in http://dicom.nema.org/medical/dicom/current/output/chtml/part04/sect_N.2.6.html). 
P.S.: Blending Mode = FOREGROUND is only for 2 inputs.

C) I did not find a **Palette Color Lookup Table Module** in **Advanced Blending Presentation State**  (while it is present in  **Blending Softcopy Presentation State CIOD**), have I missed it?

D) we can use **Threshold Type** = RANGE_INCL and have the clipping/ContrastLimit values in **Threshold Value Sequence**. See http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.11.33.html#table_C.11.33.1-1. - **OK**

E) I didn't find any relevant DICOM attribute for this.
F) This can be setup if the channel is present in the DICOM attribute **Blending Display Input Sequence**. - **OK**
NOTE: for (E) and (F), we could say that at DICOM level are the same attribute - **OK**

Punzo commented on examples/basic/index.html:
```diff
@@ -31,29 +34,38 @@
     client.searchForInstances(searchInstanceOptions).then((instances) => {
       const promises = []
       for (let i = 0; i < instances.length; i++) {
+        if (instances[i] === undefined ||
```
removed

Punzo commented on src/colorImageFrames.js:
```diff
@@ -0,0 +1,466 @@
+  const texture = generateTexture(pixelData, width, height);
+
+  const range = [...contrastLimitsRange];
+  if (bitsAllocated === 16) {
```
it is done automatically, see the varoius functions

Punzo commented on src/viewer.js:
```diff
@@ -54,7 +56,35 @@ import {
 
+// override drawTileImage method
+import {getUid} from 'ol/util.js';
+CanvasTileLayerRenderer.prototype.drawTileImage = function drawTileImage (tile, frameState, x, y, w, h, gutter, transition, opacity) {
```
I removed it, I find out that setting transition: 0, in the TileImage obejct creation is fixing the transition issue and it is not needed to overload the method

Punzo commented on src/viewer.js:
```diff
@@ -446,23 +495,608 @@ class VolumeImageViewer {
+            const channelInit = options.channelInit !== undefined ? options.channelInit.find(channelInit => channelInit.opticalPathIdentifier === opticalPathIdentifier) : undefined;
+            if (channelInit) {
+              newChannel.color = [...channelInit.color];
+              newChannel.opacity = channelInit.opacity;
```
does not relate, see my comment in https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/47#issuecomment-800323010

Punzo commented on src/viewer.js:
```diff
@@ -446,23 +495,608 @@ class VolumeImageViewer {
+    
+            const channelInit = options.channelInit !== undefined ? options.channelInit.find(channelInit => channelInit.opticalPathIdentifier === opticalPathIdentifier) : undefined;
+            if (channelInit) {
+              newChannel.color = [...channelInit.color];
```
not sure, see my https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/47#issuecomment-800323010. For me the color is a value like [255, 0, 0].

Punzo commented on src/viewer.js:
```diff
@@ -446,23 +495,608 @@ class VolumeImageViewer {
+              newChannel.opacity = channelInit.opacity;
+              newChannel.contrastLimitsRange = [...channelInit.contrastLimitsRange];
+              newChannel.visible = channelInit.visible;
+              newChannel.addToMap = channelInit.addToMap;
```
changed everything , please have again a look

Punzo commented on src/viewer.js:
```diff
@@ -446,23 +495,608 @@ class VolumeImageViewer {
+        // in the Per-Frame Functional Group Sequence, and the recipient shall not 
+        // make any assumption about the spatial position or optical path or order of the encoded frames.
+        // TO DO: get some example data.
+        console.warn('Volume Image Viewer does hot hanlde TILED_SPARSE data yet.')
```
not sure, I would need test data and I haven't them now

Punzo replied:
for the moment I have differenziate between channel and normal RGB images, so this check applies only to the channels.

Punzo commented on src/viewer.js:
```diff
@@ -446,23 +495,608 @@ class VolumeImageViewer {
+      });
+    });
+
+    if (this[_channels].length === 0) {
```
yes also the standard RGB color image are considered channel. I have updated the error message.

Punzo replied:
Update, I have differenziated between channel and normal RGB images in https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/47/commits/fffb8973b98ee79e3f48bc63a3630abed1468090

Punzo commented on src/viewer.js:
```diff
@@ -446,23 +495,608 @@ class VolumeImageViewer {
+              }
+            }
+            if (samplesPerPixel === 1) {
+              options.client.retrieveInstanceFrames(retrieveOptions).then(
```
uhm that is true, but the if the viewer has been initialized with retrieveRendered = true, in the case of channel I still need to get the data uncompressed to make blending. Also the difference before wa just that in true, it was retrievening png, if false jpeg. Now it is the same for RGB images, but for channel it actaully always retrieve as octet-stream. This can change with point (5) of the PR. So probably it would make sense to discuss this again later on.

Punzo replied:
Update, I have differenziated between channel and normal RGB images in fffb897, so there is not anymore this problem

pieper commented:
@Punzo thanks for carefully reviewing the details.  It would be good to capture the notes about the dicom blending state to a google doc that we can share with @dclunie for discussion.  We talked about providing a use case example for the standard but may very well need to see how additional features may get handled.

Punzo commented:
> @Punzo thanks for carefully reviewing the details. It would be good to capture the notes about the dicom blending state to a google doc that we can share with @dclunie for discussion. We talked about providing a use case example for the standard but may very well need to see how additional features may get handled.

there is already https://docs.google.com/document/d/1LM457qPBG1JIlwCDQQOQ1XbRC0Rz9Ea5QmN9aHny1lw/edit . Last two pages are regarding the DICOM presentation state discussion (both the blending part and the annotation one). Please let me know if there is anything unclear

Punzo commented:
Hi @hackermd, I am not sure when you will be back from vacation and you can review this PR, but I just wanted to let you know that point (2), (4) and (5) in https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/47#issue-580068584 will be implemented later on in future PRs. The issues are being tracked in the IDC project in the slim repo (https://github.com/MGHComputationalPathology/slim/projects/1, issues 11, 12 and 13).

hackermd commented on src/viewer.js:
```diff
@@ -668,154 +850,144 @@ class VolumeImageViewer {
+   * @param {number[]} thresholdValues
+   * @param {boolean} visible
+   */
+  setOpticalPathBlendingInformation(
```
I would just name this `setBlendingInformation()`

hackermd commented on src/viewer.js:
```diff
@@ -668,154 +850,144 @@ class VolumeImageViewer {
+   * @param {number[]} thresholdValues
+   * @param {boolean} visible
+   */
+  getOpticalPathBlendingInformation(opticalPathID) {
```
I would just name this `getBlendingInformation()`

hackermd commented on src/viewer.js:
```diff
@@ -668,154 +850,144 @@ class VolumeImageViewer {
-        resolutions: resolutions,
-        rotation: rotation
-      });
+    channel.blendingInformation.addToMap = true;
```
Do we need the `addToMap` attribute? Can't we just call `activateOpticalPath()` to add it to the map and `deactivateOpticalPath()` to remove it?

For the constructor, I would just add a subset of optical paths to the map. Maybe even just one to keep the initial network overhead as low as possible and thereby increase the viewer response time (time to first frames being visible to the user).

See also [this comment](https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/47/files#r612396819).

Punzo replied:
yes I removed the addToMap variable, See my comment in https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/47#discussion_r613160177 for the constructor initialization

hackermd commented on src/viewer.js:
```diff
@@ -668,154 +850,144 @@ class VolumeImageViewer {
+  /** Adds the channel to the OpenLayer Map given an id
+   * @param {string} opticalPathID of the channel
+   */
+   activateOpticalPath(opticalPathID) {
```
I would suggest also adding methods `showOpticalPath()` and `hideOpticalPath()` to allow users to toggle channel visibility without adding or removing the corresponding layers to or from the map, respectively.

hackermd commented on src/viewer.js:
```diff
@@ -668,154 +850,144 @@ class VolumeImageViewer {
+   * @param {object} channel
+   * @type {boolean} visible
+   */
+  isChannelInOpenLayerMap(channel) {
```
I suggest renaming this method to `isOpticalPathActive()`, which would better fit to `activateOpticalPath()` and `deactivateOpticalPath()`.

hackermd reviewed (changes requested):
This is fantastic work @Punzo!! I have only a few suggestions (primarily related to naming things) and reported a few bugs that I discovered while testing your branch locally.

In addition, there are a few general points that come up in multiple comments and that I wanted to discuss: 

* **Error handling**: Several methods return `null` instead of raising an error. What are the benefits and drawbacks of choosing this approach in your opinion?
* **Interface for blending information**: I would remove the `addToMap` property from `BlendingInformation` and instead rely on the `activateOpticalPath()` and `deactivateOpticalPath()` methods.
* **Tests**: It would be good if we could add a few test cases for testing the handling of multi-channel images, i.e., the handling of `metadata` by the constructor and the behavior of the `activateOpticalPath()` , `deactivateOpticalPath()`, `isOpticalPathActive()`, `getBlendingInformation()`, `setBlendingInformation()` methods

Punzo commented on examples/blend/index.html:
```diff
@@ -0,0 +1,193 @@
+    const url = 'https://server.dcmjs.org/dcm4chee-arc/aets/DCM4CHEE/rs';
+    //const url = 'http://localhost:8080/dcm4chee-arc/aets/DCM4CHEE/rs';
+    const client = new DICOMwebClient.api.DICOMwebClient({url});
+    const studyInstanceUID = '1.3.6.1.4.1.5962.99.1.2103930081.1286074986.1595536829665.3.0';
```
we have discussed this item already by email, i.e. the dcmjs had some issue with the 18 Gb dataset, we could try reupload the data (but still the perfomances were not great).  A better long-term solution would be to use a google store.

Punzo commented on src/viewer.js:
```diff
@@ -446,141 +465,297 @@ class VolumeImageViewer {
+    const layers = [];
+    if (options.blendingInformations !== undefined) {
+      this[_channels].forEach((channel) => {
+        if (channel.blendingInformation.addToMap === true) {
```
I removed the addToMap variable. For the constructor inizialiaztion (i.e. which channel add to the map), I just use the visibility. If the visibility is true, then I also add the channel layer to the map.

Punzo commented on src/viewer.js:
```diff
@@ -446,141 +465,297 @@ class VolumeImageViewer {
-      console.error('empty pyramid - no levels found')
+
+    // For blending we have to make some assumptions 
+    // 1) all channels should have the same origins, resolutions, grid sizes, tile sizes and pixel spacings (i.e. same TileGrid).
+    //    These are arrays with number of element equal to nlevel (levels of the pyramid). All channels should have the same nlevel value.
+    // 2) given (1), we calculcate the tileGrid, projection and rotation objects using the metadata of the first channel and use them for all the channels.
+    // 3) If the parameters in (1) are different, it means that we have to perfom regridding/reprojection over the data (i.e. registration).
+    //    This, at the moment, is out of scope. 
```
yes in initChannel method of the class Channel.

Punzo commented on src/viewer.js:
```diff
@@ -446,141 +465,297 @@ class VolumeImageViewer {
+      const overviewTileLayer = new TileLayer({
+        extent: this[_referenceExtents],
+        source: this[_channels][0].rasterSource,
+        preload: Infinity,
```
ok I will remove the preload, and just use the default value from Open Layer

Punzo commented on src/viewer.js:
```diff
@@ -446,141 +465,297 @@ class VolumeImageViewer {
+    }
+
+    let image = null;
+    if (this[_channels].length !== 0) {
+      image = this[_channels][0];
+    } else {
+      image = RGBimage;
+    }
```
you are right. The _initUniqueOpenLayerObjects is actually setting many instance properties. So I just moved the code into the constructor. Also for few lines did not make sense to use a function.

Punzo commented on src/viewer.js:
```diff
@@ -446,141 +465,297 @@ class VolumeImageViewer {
+      if (instanceMetadata.DimensionOrganizationType === '3D' || instanceMetadata.DimensionOrganizationType === '3D_TEMPORAL') {
+        // 3D data
+        // TO DO: get some example data.
+        console.warn('Volume Image Viewer does hot hanlde 3D data yet.')
```
I agree

Punzo commented on src/viewer.js:
```diff
@@ -446,141 +465,297 @@ class VolumeImageViewer {
+        // in the Per-Frame Functional Group Sequence, and the recipient shall not 
+        // make any assumption about the spatial position or optical path or order of the encoded frames.
+        // TO DO: get some example data.
+        console.warn('Volume Image Viewer does hot handle TILED_SPARSE ' +
```
I agree

Punzo commented on src/viewer.js:
```diff
@@ -446,141 +465,297 @@ class VolumeImageViewer {
+      if (RGBimage.OpticalPathIdentifier === '') {
+        RGBimage.OpticalPathIdentifier = instaceOpticalPathIdentifier;
+      } else if (RGBimage.OpticalPathIdentifier !== instaceOpticalPathIdentifier) {
+        console.warn('Volume Image Viewer is trying to load more than one RGB image. It is not allowed.');
```
a caller error (wrong metadata passed). I have updated the message to:

console.warn('Volume Image Viewer detected more than one color image. ' + 
                     'It is possible to load and visualize only one color image at time. ' + 
                     'Please check the input metadata. Only the first detected color image will be loaded');

Punzo commented on src/viewer.js:
```diff
@@ -446,141 +465,297 @@ class VolumeImageViewer {
+      this[_controls].overview = new OverviewMap({
+        view: overviewView,
+        layers: [overviewTileLayer],
+        collapsed: false,
```
changed back, just using openlayer default now

Punzo commented on src/viewer.js:
```diff
@@ -668,154 +850,144 @@ class VolumeImageViewer {
+   */
+   getOpticalPath(opticalPathID) {
+    if (this[_channels].length === 0) {
+      return null;
```
yes I changed to throw errors. And I also updated getOpticalPath to check for colorImage (colorImage so now is this[_colorImage])

Punzo commented on src/viewer.js:
```diff
@@ -446,141 +465,297 @@ class VolumeImageViewer {
+    
+    // We assume the first channel as the reference one for all the pyramid parameters.
+    // All the other channels have to have the same parameters.
+    this[_pyramidMetadata] = image.pyramidBaseMetadata;
```
this was just an error. I have replaced:

```
this[_pyramidMetadata] = image.pyramidBaseMetadata;
this[_metadata] = [...image.microscopyImages];

```
with 

`this[_pyramidMetadata] = [...image.pyramidMetadata];`

Punzo commented:
@hackermd I have applied the review with commit https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/47/commits/5e264f7eaa5407aea5d8e47f23116f3130d49bd1 . please have a look and let me know if there is anything else to discuss

hackermd commented on src/channel.js:
```diff
@@ -0,0 +1,554 @@
+ * @memberof channel
+ */
+
+class Channel {
```
We don't want to expose this class as part of the public API, since it exposes OpenLayers types. It is currently not included in the `dicom-microscopy-viewer.js` file, so it has not actually been exposed. However, I was wondering whether we should move it into the `viewer.js` file or rename it to `_Channel` to emphasize that the class it "private". What do you think?

Punzo replied:
renamed to _Channel

hackermd commented on src/channel.js:
```diff
@@ -0,0 +1,554 @@
+    
+    let geometryArrays = Channel.deriveImageGeometry(this);
+    
+    // Check that all the channels have the same pyramid parameters
+    if (arraysEqual(geometryArrays[0], referenceExtent) === false) {
+      throw new Error(
+        'Channels have different extent'
+      );
+    }
+    if (arraysEqual(geometryArrays[1], referenceOrigins) === false) {
+      throw new Error(
+        'Channels have different origins'
+      );
+    }
+    if (arraysEqual(geometryArrays[2], referenceResolutions) === false) {
+      throw new Error(
+        'Channels have different resolutions'
+      );
+    }
+    if (arraysEqual(geometryArrays[3], referenceGridSizes) === false) {
+      throw new Error(
+        'Channels have different grid sizes'
+      );
+    }
+    if (arraysEqual(geometryArrays[4], referenceTileSizes) === false) {
+      throw new Error(
+        'Channels have different tile sizes'
+      );
+    }
+    if (arraysEqual(geometryArrays[5], referencePixelSpacings) === false) {
+      throw new Error(
+        'Channels have different pixel spacings'
+      );
+    }
```
Let's re-formulate the error messages so that they describe the issues with the DICOM images rather than the OpenLayers objects.

hackermd replied:
The caller should not have to know anything about the OpenLayers implementation details, but should be informed about issues of the passed DICOM datasets.

Punzo replied:
ok

hackermd commented on src/channel.js:
```diff
@@ -0,0 +1,554 @@
+ *
+ * @param {object} BlendingInformation
+ * 
+ * TO DO: implement API to select focal plane in the case of 3D channels (channels with a bandwith)
+ * TO DO: use DICOM attributes for loading/saving the channel parameters (i.e. load/save the presentation state/BlendingInformation in DICOM), for example:
+ *        [x] Select area for display: http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.10.4.html 
+ *        [x] Clipping pixel values: http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.11.33.html#table_C.11.33.1-1 
+ *        [x] Select channels for display and specify the color of each channel: http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.11.34.html
+ *        [x] Blending of images: http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.11.34.html http://dicom.nema.org/medical/dicom/current/output/chtml/part04/sect_N.2.6.html
+ * TO DO: for coloring monochrome channels we download the uncompressed binaries data. 
+ *        For a large N of channels, this could be a substantial bottleneck. 
+ *        Check performance downloading binary (octet-stream) vs downloading compressed (jpeg) + decoding. 
+ *        https://github.com/cornerstonejs/codecs 
```
Shall we move some of the TODO items out of the source code and into issues?

Punzo replied:
There already issues for all these To Do in the slim repository. I removed these TO DO from the source

hackermd commented on src/channel.js:
```diff
@@ -0,0 +1,554 @@
+  }
+
+  /** Sets the channel visualization/presentation parameters
+   * @param {Object} blendingInformation : color {number[]}, 
```
Does this render correctly when using JSDoc?

Punzo replied:
fixed

hackermd reviewed (changes requested):
Thanks @punzo! I think we are almost there. There are only a few remaining things that we should address before merging the changes into master and drafting a new release.

- **API docs:** Ensure docstrings are formatted correctly to allow for automated documentation generation with JSDoc (see [parameters with properties](https://jsdoc.app/tags-param.html#parameters-with-properties) and [https://mghcomputationalpathology.github.io/dicom-microscopy-viewer](https://mghcomputationalpathology.github.io/dicom-microscopy-viewer/).
- **Coding style**: So far, we didn't specify (or enforce) any specific coding style guidelines. However, I generally try to keep lines to a max of 80 characters.
- **Test case**: It would be great if we could add at least one test case for a multi-channel image, i.e., `metadata` argument containing DICOM instances with multiple optical paths. Ideally, we would test for a few permutations and assert that the expected errors are getting raised.

Punzo commented on src/viewer.js:
```diff
@@ -402,35 +423,47 @@ const _usewebgl = Symbol('usewebgl');
       this[_usewebgl] = true;
     }
-    this[_client] = options.client;
+    this[_client] = this[_options].client;
```
I removed this[_client]  and just used  this[_options].client

Punzo commented:
> Thanks @Punzo! I think we are almost there. There are only a few remaining things that we should address before merging the changes into master and drafting a new release.
> 
> * **API docs:** Ensure docstrings are formatted correctly to allow for automated documentation generation with JSDoc (see [parameters with properties](https://jsdoc.app/tags-param.html#parameters-with-properties) and [https://mghcomputationalpathology.github.io/dicom-microscopy-viewer](https://mghcomputationalpathology.github.io/dicom-microscopy-viewer/).
> * **Coding style**: So far, we didn't specify (or enforce) any specific coding style guidelines. However, I generally try to keep lines to a max of 80 characters.
> * **Test case**: It would be great if we could add at least one test case for a multi-channel image, i.e., `metadata` argument containing DICOM instances with multiple optical paths. Ideally, we would test for a few permutations and assert that the expected errors are getting raised.

@hackermd in lasts commits I fixed the docs, I inserted new lines for long lines and I have added tests.

Punzo commented:
@hackermd with commit https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/47/commits/c244640c079479cbcc8c29de40863595fe851d13 we will use the google store set up by @pieper . Please test and let me know if we can merge this. Thanks!

Punzo commented:
@pieper @hackermd  https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/47/commits/ae656f1e7922b9eb1ecd95930ca257c0a4c4d3e4

now use png recompression. It also implements perfomances optimizations. The main one is that the offscreen render is applied only to the current tiles in view (both x and y and zoom level). Then when interacting (panning and zooming) it updates the others interaction by intercation.

hackermd commented on src/viewer.js:
```diff
@@ -446,141 +482,370 @@ class VolumeImageViewer {
+  }
+
+  _updateTilesRenderingAtPanning(){
+    if (viewer[_channels] && viewer[_channels].length !== 0) {
+      // For each channel check if any tiles at the new panning 
+      // needs to refresh the offscreen coloring rendering.
+      const tilesCoordRanges = this._transformViewCoordinatesInTilesCoordinates();
+      let render = false;
+      viewer[_channels].forEach((channel) => {  
+        const channelRender = channel.updateTilesRendering(
+          false, 
+          tilesCoordRanges[2], 
+          [tilesCoordRanges[0], tilesCoordRanges[1]]
+        );
+        if (channelRender) {
+          render = true;
+        }
+      });
+      if (render) {
+        viewer[_map].render();
```
```suggestion
    if (this[_channels] && this[_channels].length !== 0) {
      // For each channel check if any tiles at the new panning 
      // needs to refresh the offscreen coloring rendering.
      const tilesCoordRanges = this._transformViewCoordinatesInTilesCoordinates();
      let render = false;
      this[_channels].forEach((channel) => {  
        const channelRender = channel.updateTilesRendering(
          false, 
          tilesCoordRanges[2], 
          [tilesCoordRanges[0], tilesCoordRanges[1]]
        );
        if (channelRender) {
          render = true;
        }
      });
      if (render) {
        this[_map].render();
```

hackermd commented on src/renderingEngine.js:
```diff
@@ -0,0 +1,887 @@
+      pixelData,
+      decodedframeInfo, 
+      mediaType
+    } = this._checkImageTypeAndDecode(frames, false);
```
When testing locally, I am getting `this._checkImageTypeAndDecode(...) is undefined`? Did you test this with RGB images?

Punzo replied:
yes I tested with 
const studyInstanceUID = '1.2.392.200140.2.1.1.1.2.799008771.2156.1519721160.421';
const seriesInstanceUID = '1.2.392.200140.2.1.1.1.3.799008771.2156.1519721160.422';
from https://server.dcmjs.org/dcm4chee-arc/aets/DCM4CHEE/rs

note that is the options retrieveRendered is true (as default) then for RGB images the fetching order is jp2, png, jpeg. In the case of the basic example the image then is retrieve as png For png _checkImageTypeAndDecode just return undefined and in createURLFromRGBImage I just create a blob.  I added inline codes more explicit descriptions

Punzo replied:
I also just modfied the order as jp2, jpeg, png. since retrieving jpeg is faster, please let me know if ok

hackermd commented on src/viewer.js:
```diff
@@ -594,72 +859,101 @@ class VolumeImageViewer {
-          options.client.retrieveInstanceFrames(retrieveOptions).then(
+          this[_options].client.retrieveInstanceFrames(retrieveOptions).then(
             (rawFrames) => {
-              const blob = new Blob(rawFrames, {type: mediaType});
-              img.src = window.URL.createObjectURL(blob);
+              img.src = this[_renderingEngine].createURLFromRGBImage(rawFrames[0]);
             }
           );
```
This will currently fail for uncompressed images, which is an edge case but one that I already ran into.

Unfortunately, dicomweb-client currently doesn’t let us ask for “Accept: multipart/related; type=application/octet-stream, multipart/related; type=image/*”.

Would the following work?

```js
          this[_options].client.retrieveInstanceFrames(retrieveOptions).then(
            (rawFrames) => {
              img.src = this[_renderingEngine].createURLFromRGBImage(rawFrames[0]);
            }
          ).catch(
            () => {
              const retrieveOptions = {
                studyInstanceUID,
                seriesInstanceUID,
                sopInstanceUID,
                frameNumbers,
                mediaTypes: [
                  { mediaType : 'application/octet-stream' },
                ]
              };
              this[_options].client.retrieveInstanceFrames(retrieveOptions).then(
                (rawFrames) => {
                  img.src = this[_renderingEngine].createURLFromRGBImage(rawFrames[0]);
                }
              )
            }
          );

```

Punzo replied:
yes I have implemented it also for monochorme images.

hackermd reviewed (changes requested):
I ran into a few issues when testing locally with dcm4chee-arc-light

Punzo commented on examples/structured_report/index.html:
```diff
@@ -29,18 +29,22 @@
-            type="application/javascript"
-            charset="utf-8">
-    </script>
+    <script src="https://unpkg.com/@cornerstonejs/codec-libjpeg-turbo"></script>
+    <script src="https://unpkg.com/@cornerstonejs/codec-openjpeg"></script>
+    <script src="https://unpkg.com/@cornerstonejs/codec-charls"></script>
```
you are rigth, but I am having a little bit of hard time with them:

I can use import in renderingEngine.js as

`import * as libjpegturbowasm from "@cornerstonejs/codec-libjpeg-turbo/dist/libjpegturbowasm.js";
`
but then the client will look for the file `libjpegturbowasm.wasm` in the build folder. 

I will ask @dannyrb for help. For the moment I removed the decoding libraries from the direct dependencies in the package.json and I just import them in the index.html file as :

```
<script src="https://unpkg.com/@cornerstonejs/codec-libjpeg-turbo/dist/libjpegturbowasm.js"></script>
<script src="https://unpkg.com/@cornerstonejs/codec-openjpeg/dist/openjpegwasm.js"></script>
<script src="https://unpkg.com/@cornerstonejs/codec-charls/dist/charlsjs.js"></script>
```

Punzo commented:
1) @hackermd review done in https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/47/commits/891f2598bb564e888e9d19313334b02ee0101b80. 
I fixed the couple of issues that you highlighted:
A) a bug in the `_checkImageTypeAndDecode` (there was some debugger lines that I forgot to remove),
B) for `retrieveRendered` set to false, if the images are not being fetched compressed (various jpeg formats), I added the catch to donwload them as an uncompressed octree-stream.

I tested `RGB` images with the `basic` example and `monochorme` images blending with the `blend` example (both with the option `retrieveRendered` set to false and true)

2) @pieper @hackermd  regarding to study 1.3.6.1.4.1.5962.99.1.3802655970.225997140.1610120457442.3.0 (i.e. https://github.com/MGHComputationalPathology/slim/issues/7) I tried again to fetch/donwload it with dicomweb-client. I always get 406 Not Acceptable for all the frames.  the transfer sintax should be 1.2.840.10008.1.2.4.90, but I have anyway tried all of these transfer sintax with no success:
```
const jlsMediaType = 'image/jls';
const jlsTransferSyntaxUIDlossless = '1.2.840.10008.1.2.4.80';
const jlsTransferSyntaxUID = '1.2.840.10008.1.2.4.81';
const jp2MediaType = 'image/jp2';
const jp2TransferSyntaxUIDlossless = '1.2.840.10008.1.2.4.90';
const jp2TransferSyntaxUID = '1.2.840.10008.1.2.4.91';
const jpxMediaType = 'image/jpx';
const jpxTransferSyntaxUIDlossless = '1.2.840.10008.1.2.4.92';
const jpxTransferSyntaxUID = '1.2.840.10008.1.2.4.93';
const jpegMediaType = 'image/jpeg';
const jpegTransferSyntaxUID = '1.2.840.10008.1.2.4.50';

const octetStreamMediaType = 'application/octet-stream';
const octetStreamTransferSyntaxUID = '1.2.840.10008.1.2.1';
```

I tried also to open the dropbox link with latest Slicer. But it didn't work:
```
Warning in DICOM plugin Scalar Volume when examining loadable 1: Unnamed Series: Multi-frame image. If slice orientation or spacing is non-uniform then the image may be displayed incorrectly. Use with caution.
Reference image in series does not contain geometry information. Please use caution.

Loading with imageIOName: GDCM
There is more than one file, use the VectorReaderSeries instead

Algorithm vtkITKArchetypeImageSeriesVectorReaderFile(0x7a94530) returned failure for request: vtkInformation (0xe68dcd0)
  Debug: Off
  Modified Time: 321362
  Reference Count: 1
  Registered Events: (none)
  Request: REQUEST_DATA
  FROM_OUTPUT_PORT: 0
  ALGORITHM_AFTER_FORWARD: 1
  FORWARD_DIRECTION: 0


Could not read scalar volume using GDCM approach.  Error is: FileFormatError
Loading with imageIOName: DCMTK
There is more than one file, use the VectorReaderSeries instead

Algorithm vtkITKArchetypeImageSeriesVectorReaderFile(0xe0ffe80) returned failure for request: vtkInformation (0xdfdc090)
  Debug: Off
  Modified Time: 321567
  Reference Count: 1
  Registered Events: (none)
  Request: REQUEST_DATA
  FROM_OUTPUT_PORT: 0
  ALGORITHM_AFTER_FORWARD: 1
  FORWARD_DIRECTION: 0


Could not read scalar volume using DCMTK approach.  Error is: FileFormatError
Could not load: 1: Unnamed Series as a Scalar Volume

```

3) TO DO: import codecs libraries as dep.
@dannyrb At the moment the codecs library are imported in the index.html as simply:
```
<script src="https://unpkg.com/@cornerstonejs/codec-libjpeg-turbo/dist/libjpegturbowasm.js"></script>
<script src="https://unpkg.com/@cornerstonejs/codec-openjpeg/dist/openjpegwasm.js"></script>
<script src="https://unpkg.com/@cornerstonejs/codec-charls/dist/charlsjs.js"></script>
```

we should import them in the package.json:
```
"@cornerstonejs/codec-libjpeg-turbo": "^0.0.4",
"@cornerstonejs/codec-openjpeg": "^0.0.4",
"@cornerstonejs/codec-charls": "^0.0.4",
```
and then import in renderingEngine.js as:

```
import * as libjpegturbowasm from "@cornerstonejs/codec-libjpeg-turbo/dist/libjpegturbowasm.js";
import * as openjpegwasm from "@cornerstonejs/codec-openjpeg/dist/openjpegwasm.js";
import * as charlsjs from "@cornerstonejs/codec-charls/dist/charlsjs.js";
```

However:
1) then locally the client will look for the file `libjpegturbowasm.wasm` in the local `build` folder instead of in the `dist` folder of the deps.
2) the charlsjs import is undefined.

hackermd commented on src/viewer.js:
```diff
@@ -446,141 +482,387 @@ class VolumeImageViewer {
+      if (newZoom) {
+        if (Math.abs(newZoom - currZoom) > 1e-6) {
+          currZoom = newZoom;
+          viewer._updateTilesRenderingAtZoom(newZoom);
```
```suggestion
          this._updateTilesRenderingAtZoom(newZoom);
```

hackermd commented on src/viewer.js:
```diff
@@ -594,72 +876,151 @@ class VolumeImageViewer {
         } else {
-          // TODO: support "image/jp2" and "image/jls"
-          const mediaType = 'image/jpeg';
+          // allowed mediaTypes: http://dicom.nema.org/medical/dicom/current/output/chtml/part18/sect_8.7.4.html  
```
```suggestion
          console.info(`retrieve frames ${frameNumbers}`)
          // allowed mediaTypes: http://dicom.nema.org/medical/dicom/current/output/chtml/part18/sect_8.7.4.html  
```

hackermd commented on src/viewer.js:
```diff
@@ -594,72 +876,151 @@ class VolumeImageViewer {
+
+              img.src = this[_renderingEngine].createURLFromRGBImage(frameData);
+            }
+          ).catch(
```
Currently, the error will still be printed into the console. We may want to improve the error handling of the promise on the dicomweb-client side. Thoughts?

Punzo replied:
![image](https://user-images.githubusercontent.com/7985338/116862079-810ea780-ac04-11eb-862d-fc52e1a43ce0.png)
the last two errors are from dciomweb-client and we could add a flag in the retrieve methods to hide the errors (i.e. just disabling whit console.errors https://github.com/dcmjs-org/dicomweb-client/blob/beeb207ea8589461c43eb8c5e9d43c2f70d92c54/src/api.js#L152-L158). See PR https://github.com/dcmjs-org/dicomweb-client/pull/35

However we will still have this one:
![image](https://user-images.githubusercontent.com/7985338/116862644-5cff9600-ac05-11eb-9353-6c3b677ca8f2.png)
which is from the server.

hackermd commented:
@Punzo regarding the JPEG 2000 image for which requests fail, I am wondering whether that image has different subsampling factors. I had issues decoding these images with several libraries (see https://github.com/python-pillow/Pillow/issues/5165 and https://github.com/pydicom/pylibjpeg-openjpeg/issues/36).

Punzo commented:
@hackermd applied your last review in https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/47/commits/161195c9c5e5cebfcfaa38337b1f486cca6c127a . Please have a look also to https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/47#discussion_r624993846 and https://github.com/dcmjs-org/dicomweb-client/pull/35

Please note also that now this PR depends on https://github.com/dcmjs-org/dicomweb-client/pull/35



> @Punzo regarding the JPEG 2000 image for which requests fail, I am wondering whether that image has different subsampling factors. I had issues decoding these images with several libraries (see [python-pillow/Pillow#5165](https://github.com/python-pillow/Pillow/issues/5165) and [pydicom/pylibjpeg-openjpeg#36](https://github.com/pydicom/pylibjpeg-openjpeg/issues/36)).

I am not sure, I tried to load the data with dicomweb-client, 3DSlicer and Fiji, none were able to even open the data. Since I didn't generate the data, I am not sure how to proceed. It would be great if @dclunie can give some pointers what is the issue. @dclunie could youy please have a look at the point (3) in this comment https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/47#issuecomment-830008927, please?

dclunie commented:
Is the problem image one of the converted Aperio 33003 J2K images apparently created using Kakadu, which is known not to work with various J2K decoders?

Punzo commented:
ok thanks for the additional info @dclunie. I have just read that Kakadu uses Part 1, i.e. ISO/IEC 15444-1, Part 2 (coding and file format extensions), Part 3 (motion JPEG 2000), Part 9 (JPIP) and Part 15 (HTJ2K).

In this PR we use OpenJPEG for decoding JPEG2000 which can decode:
1) JPEG 2000 Image Compression (i.e. jp2, transfersintax: 1.2.840.10008.1.2.4.90 and 1.2.840.10008.1.2.4.91), 
2) JPEG 2000 Part 2 Multicomponent Image Compression (i.e. jpx, transfersintax: 1.2.840.10008.1.2.4.92 and 1.2.840.10008.1.2.4.93).

(more details at https://github.com/cornerstonejs/codecs#transfer-syntaxes)

It is true that OpenJPEG does not fully support decoding of all JPEG2000 format, but here the problem is that the frames are not even recognized as JPEG2000 from the server during the fetch/download (so we are still at the server step, not the decoding). If trying to fetch them either as:
```
const jp2MediaType = 'image/jp2';
const jp2TransferSyntaxUIDlossless = '1.2.840.10008.1.2.4.90';
const jp2TransferSyntaxUID = '1.2.840.10008.1.2.4.91';
```
or
```
const jpxMediaType = 'image/jpx';
const jpxTransferSyntaxUIDlossless = '1.2.840.10008.1.2.4.92';
const jpxTransferSyntaxUID = '1.2.840.10008.1.2.4.93';
```
The server just return error: `406 Not Acceptable` . 

Therefore, it would be nice to know exactly in which JPEG2000 Kakadu converted the images.

dclunie commented:
I don't know anything about the server, so I have no idea why it is not doing what you expect - I just provided the converted DICOM files, which are what they are, and will be "1.2.840.10008.1.2.4.91" if they were 33003 or 33005.

AFAIK, the Kakadu compressed Aperio pixel data is not using anything other than 15444-1 features, just sub-sampling that some codecs for some reason do not support, per @hackermd in a previous investigation of decoder issues related to some of my converted files.

Is this discussion in the right thread?

Punzo commented:
> I don't know anything about the server, so I have no idea why it is not doing what you expect - I just provided the converted DICOM files, which are what they are, and will be "1.2.840.10008.1.2.4.91" if they were 33003 or 33005.
> 

perfect, that is the information that I was looking for, thanks! then next step to investigate for dicomweb-client is why the server does respond with 406 Not Acceptable for "1.2.840.10008.1.2.4.91" data.

> AFAIK, the Kakadu compressed Aperio pixel data is not using anything other than 15444-1 features, just sub-sampling that some codecs for some reason do not support, per @hackermd in a previous investigation of decoder issues related to some of my converted files.
> 
> Is this discussion in the right thread?

yes let's keep the conversation only here from now on.

Punzo commented:
@hackermd rebasing into current master (with Igor's annotation PR) done. I have also added https://github.com/standard/standard config (the style is tested when running the test, i.e. npm test). Please note that there are 3 tests failing regarding to the annotations, but they are failing also in current master (i.e. release version 0.23.0). I had a look, but I am not sure, could you or @igoroctaviano have a look at the failing tests?

hackermd commented:
@Punzo the tests should be fixed with 650e33b89334769e946cc5d7102edc201df4854c. I also added a contribution guide with 8fa34a872831571bb8aed989879863c43c9a0dbb and updated the package dependencies and scripts accordingly with b4d0ecc418c43fb2cbdd95d22ccb79d9a120473b. Could you please rebase on the current `master` and try again?

Punzo commented:
> @Punzo the tests should be fixed with [650e33b](https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/commit/650e33b89334769e946cc5d7102edc201df4854c). I also added a contribution guide with [8fa34a8](https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/commit/8fa34a872831571bb8aed989879863c43c9a0dbb) and updated the package dependencies and scripts accordingly with [b4d0ecc](https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/commit/b4d0ecc418c43fb2cbdd95d22ccb79d9a120473b). Could you please rebase on the current `master` and try again?

done in https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/47/commits/bf7bcb82493ff51898d2d6688729ba52c996d4ee

hackermd reviewed (approved):
Fantastic work @Punzo!!
